src/ops/convolution.rs

```diff
@@ -0,0 +1,495 @@
+/// - weight: Weight array of shape `&[C_out, ..., C_in]`
+/// - strides: The kernel strides. All dimensions get the same stride if only one number is specified.
+/// - padding: The input padding. All dimensions get the same padding if only one number is specified.
+/// - kernel_dilation: `The  kernel dilation. All dimensions get the same dilation if only one number is specified.
+/// - input_dilation: The input dilation. All dimensions get the same dilation if only one number is specified.
+/// - groups: Input feature groups
+/// - flip: Flip the order in which the spatial dimensions of the weights are processed.
+///   Performs the cross-correlation operator when `flip` is `false` and the convolution
+///   operator otherwise.
+/// - stream: stream or device to evaluate on
+#[default_device]
+pub fn try_conv_general_device<'a>(
+    array: &Array,
+    weight: &Array,
+    strides: impl Into<Option<&'a [i32]>>,
+    padding: impl Into<Option<&'a [i32]>>,
+    kernel_dilation: impl Into<Option<&'a [i32]>>,
+    input_dilation: impl Into<Option<&'a [i32]>>,
+    groups: impl Into<Option<i32>>,
+    flip: impl Into<Option<bool>>,
+    stream: StreamOrDevice,
+) -> Result<Array, OperationError> {
+    let strides = strides.into().unwrap_or(&[1]);
+    let padding = padding.into().unwrap_or(&[0]);
+    let kernel_dilation = kernel_dilation.into().unwrap_or(&[1]);
+    let input_dilation = input_dilation.into().unwrap_or(&[1]);
+    let groups = groups.into().unwrap_or(1);
+    let flip = flip.into().unwrap_or(false);
+
+    if groups != 1 {
+        return Err(OperationError::WrongInput(
+            "groups != 1 are not supported".to_string(),
+        ));
+    }
+
+    let spatial_dims = array.ndim().checked_sub(2);
+    if spatial_dims.is_none() {
+        return Err(OperationError::WrongInput(
+            "Only 1d and 2d arrays are supported.".to_string(),
+        ));
+    }
+
+    let spatial_dims = spatial_dims.unwrap();
+
+    if spatial_dims < 1 || spatial_dims > 2 {
+        return Err(OperationError::WrongInput(
+            "Only 1d and 2d arrays are supported.".to_string(),
+        ));
+    }
+
+    if !array.dtype().is_floating() {
+        return Err(OperationError::WrongInput(
+            "Only floating point arrays are supported".to_string(),
+        ));
+    }
+
+    if array.ndim() != spatial_dims + 2 {
+        return Err(OperationError::WrongInput(
+            format!("Invalid input array with {} dimensions for {}d convolution. Expected array with {} dimensions", array.ndim(), spatial_dims, spatial_dims + 2),
+        ));
+    }
+
+    if weight.ndim() != spatial_dims + 2 {
+        return Err(OperationError::WrongInput(
+            format!("Invalid weight array with {} dimensions for {}d convolution. Expected array with {} dimensions", weight.ndim(), spatial_dims, spatial_dims + 2),
+        ));
+    }
+
+    if array.shape()[spatial_dims + 1] != weight.shape()[spatial_dims + 1] {
```

Comment:
Indexing out of bound could panic. `get()` is probably better, it returns `None` if the index is out of bound.

Reply:
I think this is safely covered by the previous checks but I've used `.get()` anyway.

Reply:
Maybe we could leave a note that this is already checked and won't panic in the future